File: bwl.Imaging/Color/RGB.vb
Public Structure RGB
    'Новых конструкторов добавлять не надо!
    Implements IRGBConvertable

    Public A As Integer
    Public R As Integer
    Public G As Integer
    Public B As Integer

    Public Sub New(r As Integer, g As Integer, b As Integer)
        Me.R = r
        Me.G = g
        Me.B = b
        Me.A = 255
    End Sub

    Public Sub New(r As Integer, g As Integer, b As Integer, a As Integer)
        Me.R = r
        Me.G = g
        Me.B = b
        Me.A = a
    End Sub

    Public Sub New(rgb As Color)
        R = rgb.R
        G = rgb.G
        B = rgb.B
        A = rgb.A
    End Sub

    Public Shared Function FromHsv(hsv As HSV) As RGB
        Return FromHsv(hsv.H, hsv.S, hsv.V, 255)
    End Function

    Public Shared Function FromHsv(h As Integer, s As Integer, v As Integer) As RGB
        Return FromHsv(h, s, v, 255)
    End Function

    Public Shared Function FromHsv(h As Integer, s As Integer, v As Integer, alpha As Integer) As RGB
        Dim valuemax = 255
        Dim rgb As RGB
        Dim hi = Math.Floor(h / 60.0) Mod 6
        Dim vmin = (valuemax - s) * v / valuemax
        Dim a = (v - vmin) * ((h Mod 60) / 60.0)
        Dim vinc = vmin + a
        Dim vdec = v - a
        Select Case hi
            Case 0 : rgb.R = v : rgb.G = vinc : rgb.B = vmin
            Case 1 : rgb.R = vdec : rgb.G = v : rgb.B = vmin
            Case 2 : rgb.R = vmin : rgb.G = v : rgb.B = vinc
            Case 3 : rgb.R = vmin : rgb.G = vdec : rgb.B = v
            Case 4 : rgb.R = vinc : rgb.G = vmin : rgb.B = v
            Case 5 : rgb.R = v : rgb.G = vmin : rgb.B = vdec
        End Select
        rgb.A = alpha
        Return rgb
    End Function

    Public Function ToHSV() As HSV
        Return HSV.FromRgb(Me)
    End Function

    Public Function ToColor() As Color
        Return Color.FromArgb(Limit(A), Limit(R), Limit(G), Limit(B))
    End Function

    Public Shared Function Limit(data As Integer) As Byte
        If data < 0 Then data = 0
        If data > 255 Then data = 255
        Return data
    End Function

    Public Function ToRGB() As RGB Implements IRGBConvertable.ToRGB
        Return Me
    End Function
End Structure

File: Bwl.Imaging.Skia/Color/RGB.cs


using SkiaSharp;

namespace Bwl.Imaging.Skia
{

    // Новых конструкторов добавлять не надо!
    public struct RGB : IRGBConvertable
    {

        public int A;
        public int R;
        public int G;
        public int B;

        public RGB(int r, int g, int b)
        {
            R = r;
            G = g;
            B = b;
            A = 255;
        }

        public RGB(int r, int g, int b, int a)
        {
            R = r;
            G = g;
            B = b;
            A = a;
        }

        public RGB(SKColor rgb)
        {
            R = rgb.Red;
            G = rgb.Green;
            B = rgb.Blue;
            A = rgb.Alpha;
        }

        public static RGB FromHsv(HSV hsv)
        {
            return FromHsv(hsv.H, hsv.S, hsv.V, 255);
        }

        public static RGB FromHsv(int h, int s, int v)
        {
            return FromHsv(h, s, v, 255);
        }

        public static RGB FromHsv(int h, int s, int v, int alpha)
        {
            int valuemax = 255;
            var rgb = default(RGB);
            int hi = ImagingMath.Limit(h / 60.0d) % 6;
            byte vmin = ImagingMath.Limit((valuemax - s) * v / (double)valuemax);
            double a = (v - vmin) * (h % 60 / 60.0d);
            byte vinc = ImagingMath.Limit(vmin + a);
            byte vdec = ImagingMath.Limit(v - a);
            switch (hi)
            {
                case 0:
                    {
                        rgb.R = v;
                        rgb.G = vinc;
                        rgb.B = vmin;
                        break;
                    }
                case 1:
                    {
                        rgb.R = vdec;
                        rgb.G = v;
                        rgb.B = vmin;
                        break;
                    }
                case 2:
                    {
                        rgb.R = vmin;
                        rgb.G = v;
                        rgb.B = vinc;
                        break;
                    }
                case 3:
                    {
                        rgb.R = vmin;
                        rgb.G = vdec;
                        rgb.B = v;
                        break;
                    }
                case 4:
                    {
                        rgb.R = vinc;
                        rgb.G = vmin;
                        rgb.B = v;
                        break;
                    }
                case 5:
                    {
                        rgb.R = v;
                        rgb.G = vmin;
                        rgb.B = vdec;
                        break;
                    }
            }
            rgb.A = alpha;
            return rgb;
        }

        public HSV ToHSV()
        {
            return HSV.FromRgb(this);
        }

        public SKColor ToColor()
        {
            return new SKColor(ImagingMath.Limit(R), ImagingMath.Limit(G), ImagingMath.Limit(B), ImagingMath.Limit(A));
        }

        public RGB ToRGB()
        {
            return this;
        }
    }
}